### PR TITLE
[WIP] Feature/ppcdev 8176 step by step guide

### PR DESCRIPTION
**Pull request by plks**

### CODE REVIEW

File: platform/authorized_api_guide.md [hunk 1]
```diff
@@ -38,22 +73,247 @@ curl 'https://<domain>/auth/token' -H "Content-Type: application/json" --data '{
...
+    },
+    "type": "ppms/app",
+    "id": "b30e538d-4b05-4a75-ae25-7eb565901f38"
+  }
+}'
+```
+
+This request changed app name from `AppName` to `NewAppName`. Notice three things:
+* `-X PATCH` before URL. It means that this request is available using `HTTP PATCH method`
+* you have to specify also `data/id` - it's a [jsonapi]((http://jsonapi.org/)) requirement
```
Comment (mrfijas) [inline platform/authorized_api_guide.md]: jsonapi link is broken (to much `()`)

I wonder what's correct spelling of jsonapi.. should it be `JsonApi`, `jsonApi`, `{json:api}`, `Json API`? :derp:

File: platform/authorized_api_guide.md [hunk 2]
```diff
@@ -38,22 +73,247 @@ curl 'https://<domain>/auth/token' -H "Content-Type: application/json" --data '{
...
+* also `data/type` is important. For example, when you want to work with app resource, specify it's type as `ppms/app` 
+
+There is no response example. API will return `204 No Content` status code.
```
Comment (mrfijas) [inline platform/authorized_api_guide.md]: maybe "API will return `204 No Content` status code with an empty response"?

File: platform/authorized_api_guide.md [hunk 3]
```diff
@@ -38,22 +73,247 @@ curl 'https://<domain>/auth/token' -H "Content-Type: application/json" --data '{
...
+
+
+#### Delete an app
+
+Sometimes, resources are not needed anymore so lets look an example on how to delete them.
```
Comment (mrfijas) [inline platform/authorized_api_guide.md]: "let's have a look at example" or something like that

File: platform/authorized_api_guide.md [hunk 4]
```diff
@@ -38,22 +73,247 @@ curl 'https://<domain>/auth/token' -H "Content-Type: application/json" --data '{
...
+
+```
+DELETE /api/apps/v2/{appId}
+```
+
+Request example:
+```
+curl -X DELETE 'https://<domain>/api/apps/v2/b30e538d-4b05-4a75-ae25-7eb565901f38'  -H "Authorization: Bearer <your_access_token>" -H "Content-Type: application/vnd.api+json"
```
Comment (mrfijas) [inline platform/authorized_api_guide.md]: here, and before there are two spaces before `-H` flag

File: platform/authorized_api_guide.md [hunk 5]
```diff
@@ -38,22 +73,247 @@ curl 'https://<domain>/auth/token' -H "Content-Type: application/json" --data '{
...
+There is no response example. API will return `204 No Content` status code.
+
+### API usage example with Postman
+
+[Postman](https://www.getpostman.com/) is a multiplatform GUI application for creating API calls.
+PPMS allows you to export swagger documentation and easily import it to Postman.
+Depending of what you want to work with, you can import given swagger docs:
+* [Apps](https://raw.githubusercontent.com/PiwikPRO/PPMS-PublicDocs/master/platform/authorized_api/apps/index.json)
+* [Users](https://raw.githubusercontent.com/PiwikPRO/PPMS-PublicDocs/master/platform/authorized_api/users/index.json)
+* [Access control](https://github.com/PiwikPRO/PPMS-PublicDocs/blob/master/platform/authorized_api/access_control/index.json)
```
Comment (mrfijas) [inline platform/authorized_api_guide.md]: it's not `raw.gith..`

File: platform/authorized_api_guide.md [hunk 6]
```diff
@@ -38,22 +73,247 @@ curl 'https://<domain>/auth/token' -H "Content-Type: application/json" --data '{
...
+
+Simply click in Postman: `import -> Import From Link`. Then, all of your paths are imported!
+You have to override two things:
+* replace your domain in url
+* add token. Click on `Authorization` tab on chosen API call and then use Bearer Token type. 
+  Paste your token, and now, you can call API using `SEND` button.
+
+## FAQ
```
Comment (mrfijas) [inline platform/authorized_api_guide.md]: 👍 bit thumbsup for this.

maybe it could be broaden with `{ "error": "access token not authorized" }` (on invalid token)

File: platform/authorized_api_guide.md [hunk 7]
```diff
@@ -20,15 +21,49 @@ API credentials which then allows you to request for a token that is used for au
...
 
-### Create access token
+#### Create access token
 
 Having generated your API Credentials, now you are ready for creating access token that will be used in communication with API.
-To do so, send POST request 
+
+Piwik PRO API tokens use [jwt](https://jwt.io/) format.
+
+Now, you can use obtained `access_token` for communication with Piwik PRO API.
```
[the remark this answers is not in the document]
Comment (plks) [reply platform/authorized_api_guide.md]: yes, my mistake

File: platform/authorized_api_guide.md [hunk 8]
```diff
@@ -20,15 +21,49 @@ API credentials which then allows you to request for a token that is used for au
...
+For sake of this examples, `https://<domain>` is a URL of your PPMS instance (e.g. `https://example.piwik.pro`) and our goal
+is to perform CRUD operations on an app. We will as follow:
+* create an app
+* read created app
+* update some fields
+* in the end, we will remove given app
+
 
-Curl example:
+First step is needed only once, for generating of access token.
```
[the remark this answers is not in the document]
Comment (plks) [reply platform/authorized_api_guide.md]: proposal: `Once you generated an access token, you can use it as long as it is valid.`

File: platform/authorized_api_guide.md [hunk 9]
```diff
@@ -38,22 +73,247 @@ curl 'https://<domain>/auth/token' -H "Content-Type: application/json" --data '{
...
+
+Note, that you have to replace: 
+* `<domain>` with your PPMS instance URL,
+* `<your_access_token>` with your generated access token 
+
+Response example:
+```
+{
+   "data":{
+      "type":"ppms\/app",
```
[the remark this answers is not in the document]
Comment (plks) [reply platform/authorized_api_guide.md]: Yes, I wanted to make it consistent

File: platform/authorized_api_guide.md [hunk 10]
```diff
@@ -38,22 +73,247 @@ curl 'https://<domain>/auth/token' -H "Content-Type: application/json" --data '{
...
 ```
 
-The example response is: 
+Response example:
 ```
 {"token_type":"Bearer","expires_in":86400,"access_token":"your_access_token"}
 ```
 
-Piwik PRO API tokens use [jwt](https://jwt.io/) format.
+Field `access_token` contains your token which then will be used for all API calls.
```
[the remark this answers is not in the document]
Comment (plks) [reply platform/authorized_api_guide.md]: I'm not sure. We have written it is JWT, we have written it is a token. It could be easily googleable. I'm not sure what technical knowledge of end users will be.

File: platform/authorized_api_guide.md [hunk 11]
```diff
@@ -38,22 +73,247 @@ curl 'https://<domain>/auth/token' -H "Content-Type: application/json" --data '{
...
+#### Read an app
+
+Now, when app is added, it is possible to read it.
+
+Request example:
+```
+GET /api/apps/v2/{appId}
+```
+
+Request example:
```
[the remark this answers is not in the document]
Comment (plks) [reply platform/authorized_api_guide.md]: Why? I also want it to make consistent. CRUD request and response examples. Whatever you need, you can copy, apply your domain, id, and paste.

File: platform/authorized_api_guide.md (hunk 1 above)
Comment (plks) [reply platform/authorized_api_guide.md]: first line of json api docs `If you’ve ever argued with your team about the way your JSON responses should be formatted, JSON API can be your anti-bikeshedding tool.`. I'm moving it to JSON API across docs. Good point

File: platform/authorized_api_guide.md (hunk 4 above)
Comment (plks) [reply platform/authorized_api_guide.md]: :eye:

File: platform/authorized_api/users/index.json [hunk 12]
```diff
@@ -202,6 +234,9 @@
     },
     "Not_Found": {
       "description": "Not Found"
+    },
+    "No_Content": {
+      "description": "No content"
```
[the remark this answers is not in the document]
Comment (plks) [reply platform/authorized_api/users/index.json]: but it is autogenerated file. I haven't created it. We could write other json linter rules.

File: platform/authorized_api_guide.md [hunk 13]
```diff
@@ -20,15 +21,49 @@ API credentials which then allows you to request for a token that is used for au
...
+
 ```
-POST /auth/token
+Authorization: Bearer your-token-here
 ```
+  
+### Deleting API Credentials
+
+Once you want to revoke the possibility of generating API token using given `CLIENT ID` and `CLIENT SECRET`,
+go to to `My profile` and click `Delete` button on selected API credentials.
```
[the remark this answers is not in the document]
Comment (plks) [reply platform/authorized_api_guide.md]: :eye: :eyes:  :clap:

File: platform/authorized_api_guide.md (hunk 8 above)
[the remark this answers is not in the document]
Comment (mrfijas) [reply platform/authorized_api_guide.md]: `Once you generated an access token, you can use it during its lifetime (xx hours by default)`?.
> https://www.oauth.com/oauth2-servers/access-tokens/access-token-lifetime/

File: platform/authorized_api_guide.md (hunk 10 above)
[the remark this answers is not in the document]
Comment (mrfijas) [reply platform/authorized_api_guide.md]: in layman's terms token is "d1389hf31298fh1" or "18417", but not long elegy of weird symbols and dots. maybe we could show real JWT in example response from oauth?

File: platform/authorized_api_guide.md (hunk 11 above)
[the remark this answers is not in the document]
Comment (mrfijas) [reply platform/authorized_api_guide.md]: yeah, but here it's twice: before and between `GET ..` and `curl` example - in other cases theres only one "request example" text

File: platform/authorized_api_guide.md [hunk 14]
```diff
@@ -38,22 +73,247 @@ curl 'https://<domain>/auth/token' -H "Content-Type: application/json" --data '{
...
+GET /api/apps/v2/{appId}
+```
+
+Request example:
+```
+curl 'https://testing.piwik.pro/api/apps/v2/b30e538d-4b05-4a75-ae25-7eb565901f38'  -H "Authorization: Bearer <your_access_token>" -H "Content-Type: application/vnd.api+json"
+```
+
+Notice:
+* URL contains `b30e538d-4b05-4a75-ae25-7eb565901f38`. What is it? It is unique ID of an app.
```
[the remark this answers is not in the document]
Comment (mrfijas) [reply platform/authorized_api_guide.md]: maybe it would be nice to wrap these notices in `>`? so it would be formatted like:
> Notice:
> 
> URL contains `b30e538d-4b05-4a75-ae25-7eb565901f38`. What is it? It is unique ID of an app.

File: platform/authorized_api_guide.md (hunk 10 above)
[the remark this answers is not in the document]
Comment (plks) [reply platform/authorized_api_guide.md]: it would not hurt anybody, but it could only help

File: platform/authorized_api_guide.md (hunk 11 above)
[the remark this answers is not in the document]
Comment (plks) [reply platform/authorized_api_guide.md]: ahh ofc you are right. It was about text `Request example`, I thought you mean this section with curl example

File: platform/authorized_api_guide.md [hunk 15]
```diff
@@ -38,22 +73,247 @@ curl 'https://<domain>/auth/token' -H "Content-Type: application/json" --data '{
 }' 
 ```
 
-The example response is: 
+Response example:
 ```
 {"token_type":"Bearer","expires_in":86400,"access_token":"your_access_token"}
```
[the remark this answers is not in the document]
Comment (plks) [reply platform/authorized_api_guide.md]: but I reformated because of pasting long real jwt token

File: platform/authorized_api_guide.md [hunk 16]
```diff
@@ -38,22 +73,247 @@ curl 'https://<domain>/auth/token' -H "Content-Type: application/json" --data '{
...
+         ],
+         "organization":"default",
+         "appType":"web",
+         "gdprLocationRecognition":false
+      }
+   }
+}
+
+```
+#### Read an app
```
[the remark this answers is not in the document]
Comment (plks) [reply platform/authorized_api_guide.md]: Get + CRUD info removed. I thought that this is documentation of API not ORM library

File: platform/authorized_api/apps/index.json [hunk 17]
```diff
@@ -385,6 +402,12 @@
             "*.example.com",
             "example.com"
           ]
+        },
+        "gdprLocationRecognition": {
+          "type": "boolean",
+          "description": "Use geoIP to detect UE visitors and present consent forms only to them",
```
Comment (rezik007) [inline platform/authorized_api/apps/index.json]: We fixed this description a little bit so here should be `geolocation` instead of `geoIP`

File: platform/authorized_api_guide.md [hunk 18]
```diff
@@ -20,15 +21,57 @@ API credentials which then allows you to request for a token that is used for au
...
+{"token_type":"Bearer","expires_in":86400,"access_token":"<your_access_token>"}
 ```
 
-Curl example:
+Now, you can use obtained `<access_token>` for communication with Piwik PRO API.
+Field `expires_in` stands for time (in seconds) for token expiration (TTL).
+As token is a Bearer type, it must be **included in every API call** within header.
+
+```
+Authorization: Bearer <your-token-here>
```
Comment (mrfijas) [inline platform/authorized_api_guide.md]: to be consistent: `<access_token>` (we've got `access_token`, `your_access_token` and `token` - in FAQ)?

Comment (mrfijas) [reply platform/authorized_api_guide.md]: also maybe we could stick to one naming convention? we've got snake and camel cases here (`clientId` vs `access_token`)

File: platform/authorized_api_guide.md [hunk 19]
```diff
@@ -20,15 +21,49 @@ API credentials which then allows you to request for a token that is used for au
...
-Curl example:
+First step is needed only once, for generating of access token.
+
+#### Generate your access token
+
+Request example:
+
+```
+POST /auth/token
+```
```
[the remark this answers is not in the document]
Comment (mrfijas) [reply platform/authorized_api_guide.md]: I think it's better to do it explicitly and not to depend on curls magic that on `curl example.org` it does `GET`, and on `curl -d 'some body' example.org` does `POST`

File: platform/authorized_api/apps/index.json [hunk 20]
```diff
@@ -46,6 +50,16 @@
       "description": "Organization to which an app belongs to",
       "example": "default"
     },
+    "App_Type": {
```
[the remark this answers is not in the document]
Comment (RomanSek) [reply platform/authorized_api/apps/index.json]: Master is for latest version that was deployed, so it's 6.2 currently.

File: platform/authorized_api_guide.md [hunk 21]
```diff
@@ -20,15 +21,57 @@ API credentials which then allows you to request for a token that is used for au
...
+  
+### Deleting API Credentials
+
+Once you want to revoke the possibility of generating API token using given `CLIENT ID` and `CLIENT SECRET`,
+go to `My profile` and click `Delete` button on selected API credentials.
+
+## API usage example
+
+Whatever API call you choose, first remember that you must generate
```
Comment (rezik007) [inline platform/authorized_api_guide.md]: I rather understand that this way: I have previously generated credentials and I HAD token generated week ago but after this time I want to perform some requests to API... so this section shouldn't include information about:
1. first - have generated new (or old but saved somewhere) credentials
2. second - generate a FRESH token (if expired but generating new one could be a preferable way) required for every single request
???

File: platform/authorized_api_guide.md (hunk 19 above)
[the remark this answers is not in the document]
Comment (rezik007) [reply platform/authorized_api_guide.md]: @plks so maybe make a note about it? or write two curl examples that are equal?

File: platform/authorized_api_guide.md [hunk 22]
```diff
@@ -38,22 +81,255 @@ curl 'https://<domain>/auth/token' -H "Content-Type: application/json" --data '{
...
+Request example:
+```
+PATCH /api/apps/v2/<appId>
+```
+
+```
+curl -X PATCH 'https://<domain>/api/apps/v2/b30e538d-4b05-4a75-ae25-7eb565901f38' -H "Authorization: Bearer <your_access_token>" -H "Content-Type: application/vnd.api+json" -v --data '{
+  "data": {
+    "attributes": {
+      "name": "NewAppName"
```
Comment (mrfijas) [inline platform/authorized_api_guide.md]: hmm.. now I think that it's worth mentioning somewhere what attributes we can change - everything but `id` and `appType` and that all these attributes are optional (so you provide only ones you want to change)

File: platform/authorized_api_guide.md [hunk 23]
```diff
@@ -38,22 +81,255 @@ curl 'https://<domain>/auth/token' -H "Content-Type: application/json" --data '{
...
+         "excludedIps":[
+
+         ],
+         "excludedUrlParams":[
+
+         ],
+         "excludedUserAgents":[
+
+         ],
+         "gdpr":true,
```
Comment (rezik007) [inline platform/authorized_api_guide.md]: Asking just to be sure: Those fields are always returned even when Consent Manager is not available?

File: platform/authorized_api_guide.md [hunk 24]
```diff
@@ -38,22 +81,255 @@ curl 'https://<domain>/auth/token' -H "Content-Type: application/json" --data '{
...
+}'
+```
+
+This request changed app name from `AppName` to `NewAppName`. 
+> Notice three things:
+> * `-X PATCH` before URL. It means that this request is available using `HTTP PATCH method`
+> * you have to specify also `data/id` - it's a [JSON API](http://jsonapi.org/) requirement
+> * also `data/type` is important. For example, when you want to work with app resource, specify it's type as `ppms/app` 
+
+API will return `204 No Content` status code with an empty response
```
Comment (rezik007) [inline platform/authorized_api_guide.md]: Missing dot :)

File: platform/authorized_api_guide.md [hunk 25]
```diff
@@ -38,22 +81,255 @@ curl 'https://<domain>/auth/token' -H "Content-Type: application/json" --data '{
...
+    "id": "b30e538d-4b05-4a75-ae25-7eb565901f38"
+  }
+}'
+```
+
+This request changed app name from `AppName` to `NewAppName`. 
+> Notice three things:
+> * `-X PATCH` before URL. It means that this request is available using `HTTP PATCH method`
+> * you have to specify also `data/id` - it's a [JSON API](http://jsonapi.org/) requirement
+> * also `data/type` is important. For example, when you want to work with app resource, specify it's type as `ppms/app`
```
Comment (rezik007) [inline platform/authorized_api_guide.md]: `is important and also required`?

File: platform/authorized_api_guide.md [hunk 26]
```diff
@@ -38,22 +81,255 @@ curl 'https://<domain>/auth/token' -H "Content-Type: application/json" --data '{
...
+
+Remember, all API calls needs to be created with `Content-Type: application/vnd.api+json` header. 
+If you use `curl` you need to use `-H "Content-Type: application/json"` flag. 
+Postman allows configuring headers with `Header` tab.
+
+### API returns `JWT not found`
+
+Remember, you need to always use your API token. You need to send it all the time within `Authorization: Bearer <token>` header.
+If you use `curl` you need to use `-H "Authorization: Bearer <token>"` flag.
+Postman allows configuring tokens in authorization tab. Choose type `Bearer Token` and paste it here.
```
Comment (rezik007) [inline platform/authorized_api_guide.md]: `here` -> `there`?

File: platform/authorized_api_guide.md [hunk 27]
```diff
@@ -38,22 +81,255 @@ curl 'https://<domain>/auth/token' -H "Content-Type: application/json" --data '{
...
+Remember, all API calls needs to be created with `Content-Type: application/vnd.api+json` header. 
+If you use `curl` you need to use `-H "Content-Type: application/json"` flag. 
+Postman allows configuring headers with `Header` tab.
+
+### API returns `JWT not found`
+
+Remember, you need to always use your API token. You need to send it all the time within `Authorization: Bearer <token>` header.
+If you use `curl` you need to use `-H "Authorization: Bearer <token>"` flag.
+Postman allows configuring tokens in authorization tab. Choose type `Bearer Token` and paste it here. 
+Remember to keeping this token secure as it allows access to sensitive data!
```
Comment (rezik007) [inline platform/authorized_api_guide.md]: `keeping`->`keep`?

File: platform/authorized_api_guide.md [hunk 28]
```diff
@@ -38,22 +81,255 @@ curl 'https://<domain>/auth/token' -H "Content-Type: application/json" --data '{
...
+
+### API returns `Expired JWT Token`
+
+Every token that you generated is specified by TTL - time to live. By default it's 24h.
+After token is expired, you need to [generate your access token](#generate-your-access-token)
+
+### API returns `access token not authorized`
+
+This message means, that you sent access token within proper `Authorization: Bearer` field, although it is invalid.
+Make sure
```
Comment (rezik007) [inline platform/authorized_api_guide.md]: `Make sure`... and? :)

File: platform/authorized_api_guide.md (hunk 18 above)
Comment (plks) [reply platform/authorized_api_guide.md]: Rewritten to `your_access_token`, camel case reduces as much as possible

File: platform/authorized_api_guide.md (hunk 19 above)
[the remark this answers is not in the document]
Comment (plks) [reply platform/authorized_api_guide.md]: ok, some developers even add this flag for no reason when using `-d` flag

File: platform/authorized_api_guide.md (hunk 22 above)
Comment (plks) [reply platform/authorized_api_guide.md]: `you can set only parameters you want to update. For more apps attributes go to [App edit reference](https://developers.piwik.pro/en/latest/platform/authorized_api/apps/apps_api.html#operation/api_app_edit_v2)`

Comment (mrfijas) [reply platform/authorized_api_guide.md]: sorry, it was just a thought - but probably more targeted at docs of apps (https://developers.piwik.pro/en/latest/platform/authorized_api/apps/apps_api.html#operation/api_app_edit_v2), not this generic "api guide"

File: platform/authorized_api_guide.md (hunk 21 above)
Comment (plks) [reply platform/authorized_api_guide.md]: as we talk offline I will try to be more precise with generation of token & api credentials

File: platform/authorized_api_guide.md (hunk 23 above)
Comment (plks) [reply platform/authorized_api_guide.md]: yes they are, but possibly that's ok

File: platform/authorized_api_guide.md (hunk 24 above)
Comment (plks) [reply platform/authorized_api_guide.md]: :eyes:

File: platform/authorized_api_guide.md (hunk 25 above)
Comment (plks) [reply platform/authorized_api_guide.md]: I left only `required`. It should be clear that this way it is important

File: platform/authorized_api_guide.md (hunk 28 above)
Comment (plks) [reply platform/authorized_api_guide.md]: cry

File: platform/authorized_api_guide.md (hunk 23 above)
Comment (mrfijas) [reply platform/authorized_api_guide.md]: yup - we'll come back to this problem after implementing proper licensing - modifying API response because of product-bar config seems wrong
